base/c9s-python-3.11/.gitignore

```diff
@@ -0,0 +1,6 @@
+# pipenv
+#   According to pypa/pipenv#598, it is recommended to include Pipfile.lock in version control.
+#   However, in case of collaboration, if having platform-specific dependencies or dependencies
+#   having no cross-platform support, pipenv may install dependencies that don't work, or not
+#   install all needed dependencies.
+Pipfile.lock
```

Comment:
Is there a reason we don't use one global .gitignore file and having this in multiple directories?

Also, I'm not sure I truly like it - it's in the repo but we put it into ignore... maybe I would rather have something like pre-commit or GHA that checks that the Pipfile.lock isn't changed by user? :thinking: but maybe even that no... we should simply tell that the PR is bad if it is updating the Pipfile.lock instead of using this new GHA, no? In other words - why having this file in gitignore? :thinking: :shrug:

Reply:
>Is there a reason we don't use one global .gitignore file and having this in multiple directories?

Yes, this way, if we deprecate a notebook flavor will deprecate also this restriction, the same if we add a new notebook.  On the other side, on the centralized .gitignore file we need to keep maintain the paths of the notebooks  that is applies. 

>why having this file in gitignore?
To avoid different variations of Pipfile.lock that generated the users local envs while they push to the repo and cause conflicts. 
Added a section on the readme file and on the contributing.md: 

**Pipfile.lock Generation**
To maintain **consistency across different local environments** and prevent platform-specific issues, users should not push their Pipfile.lock files after local execution. Collaborators may encounter discrepancies due to varying local setups, leading to potential conflicts during the lock file generation. Instead, users can update these files through the [piplock-renewal.yaml](https://github.com/opendatahub-io/notebooks/blob/main/.github/workflows/piplock-renewal.yaml) GitHub Action. 

This action allows users to specify a branch for updating and automerging the Pipfile.lock files, select the desired Python version, and choose whether to include optional directories in the update.
*NOTE:* To ensure the GitHub Action runs successfully, users must create a `GH_ACCESS_TOKEN` secret in their fork.

Reply:
> Yes, this way, if we deprecate a notebook flavor will deprecate also this restriction, the same if we add a new notebook. 

I don't follow. What does it mean when we deprecate? When we don't want this restriction to be in place anymore?

> To avoid different variations of Pipfile.lock that generated the users local envs while they push to the repo and cause conflicts.

But people don't push to the repo usually. We use PRs for this. And we can tell user to revert such changes in their PR. Also, I don't know in which situation user has different content of the Pipfile.lock files than what is in the repo. Usually only in case of development, no? TBH, I don't feel comfortable to seeing different results when I build image locally compared to when it is done for us via openshift-ci/GHA and pushed to quay... having these files in gitignore will mask me the information that I have some local changes on the repository that will result in different image to what is on the upstream repo with the very same commit e.g.

Reply:
>> Is there a reason we don't use one global .gitignore file and having this in multiple directories?

>Yes, this way, if we deprecate a notebook flavor will deprecate also this restriction, the same if we add a new notebook. On the other side, on the centralized .gitignore file we need to keep maintain the paths of the notebooks that is applies.

add a top level .gitignore and put there

```
**/Pipfile.lock
```

problem solved

> Two consecutive asterisks ("**") in patterns matched against full pathname may have special meaning:
>
> A leading "**" followed by a slash means match in all directories.
>
> https://git-scm.com/docs/gitignore#_pattern_format

Reply:
When we deprecate a feature, it means we mark it as outdated or no longer recommended for use, with the intention of removing it in the future. In this specific case, if we deprecate a notebook flavor, it means we no longer actively support it. Consequently, any restrictions associated with it—such as this one—would also be deprecated. Similarly, if we add a new notebook flavor, relevant restrictions would naturally apply to that new flavor.


And about your concerns about masking differences and ensuring consistency in generated images are noted, but let me address the key issues:

We should rely on a single, consistent source for the Pipfile.lock generation to avoid discrepancies. The images generated by GHA (GitHub Actions) or OpenShift CI (ocp-ci) serve this purpose because these are the images ultimately delivered to customers. This way, we can ensure that all builds, whether local or in CI, align with the same source of truth.

For users, the process is straightforward:

- Trigger the GHA to generate the Pipfile.lock files for their branch and Python version.
- Fetch the generated files back with a simple git pull.

This eliminates any masking or differing results between local and upstream, providing complete transparency and consistency.

The main objective of this PR is to address conflicts arising between upstream and downstream branches, as detailed in the Jira issue. These conflicts are caused by OS-specific differences (e.g., markers in Pipfile.lock) that occur when Pipfile.lock is generated across various local environments. Such discrepancies lead to unnecessary diffs and merge conflicts, as observed in PRs like [#423](https://github.com/red-hat-data-services/notebooks/pull/423) and [#424](https://github.com/red-hat-data-services/notebooks/pull/424#issuecomment-2453039347). Centralizing this process resolves these issues by standardizing the generated files.

If there is a more effective way to eliminate these conflicts while maintaining consistency across environments and CI, we are open to hearing it. Until then, this approach ensures a clean, reliable, and unified process.

Reply:
> having these files in gitignore will mask me the information that I have some local changes on the repository that will result in different image to what is on the upstream repo with the very same commit e.g.

isn't that the same problem as if you forget commit some new files and then you wonder why locally you get different results than what ci gives you?

Reply:
> add a top level .gitignore and put there

yeah, exactly what I meant - If we really need it then have it globally on one place and not on multiple places... then I didn't understand the note about the deprecation, so I'm missing something here.

> Consequently, any restrictions associated with it—such as this one—would also be deprecated. Similarly, if we add a new notebook flavor, relevant restrictions would naturally apply to that new flavor.

Right, but I still don't understand how is this related to the .gitignore and Pipfile.lock files.

> We should rely on a single, consistent source for the Pipfile.lock generation to avoid discrepancies. The images generated by GHA (GitHub Actions) or OpenShift CI (ocp-ci) serve this purpose because these are the images ultimately delivered to customers. This way, we can ensure that all builds, whether local or in CI, align with the same source of truth.

Definitely agree, that is what I wanted in the past already - having a GHA to update the Pipfile.lock files for us with a unified way. So this part is good from my side and I never objected it.

> These conflicts are caused by OS-specific differences (e.g., markers in Pipfile.lock) that occur when Pipfile.lock is generated across various local environments. Such discrepancies lead to unnecessary diffs and merge conflicts, as observed in PRs like https://github.com/red-hat-data-services/notebooks/pull/423 and https://github.com/red-hat-data-services/notebooks/pull/424#issuecomment-2453039347. Centralizing this process resolves these issues by standardizing the generated files.

Right, but still - do we really want to have this in .gitignore since it's part of the repository anyway? As I wrote above - people make changes to the repository via PRs. We control what goes in via PR. We can say that the change in pipfile.lock need to be removed from the PR...

> isn't that the same problem as if you forget commit some new files and then you wonder why locally you get different results than what ci gives you?

Okay, so I tried locally now. I thought that the change in the file won't be shown in `git diff` command. But it isn't true - it actually is. Also, looks like if the Pipfile.lock is already part of the repository, then the .gitignore doesn't work for it for me? I can add the file even without the `-f` option and commit it... what am I doing wrong? If I understand correctly - us adding these into .gitignore and still having them in the repo actually doesn't change anything from the user POV? :thinking:

Reply:
> Okay, so I tried locally now. I thought that the change in the file won't be shown in git diff command. But it isn't true - it actually is. Also, looks like if the Pipfile.lock is already part of the repository, then the .gitignore doesn't work for it for me? I can add the file even without the -f option and commit it... what am I doing wrong? If I understand correctly - us adding these into .gitignore and still having them in the repo actually doesn't change anything from the user POV? 🤔

Okay, so to stop tracking - I should probably do this: https://stackoverflow.com/a/18277622 (haven't tried that yet). But then, my original concern is back - I won't see my local changes of the file against to the content in the repo so I will be confused why my local image build is different to what is in quay.io....

Reply:
> If there is a more effective way to eliminate these conflicts while maintaining consistency across environments and CI, we are open to hearing it.

I think I tried to propose someting already :slightly_smiling_face: : 

> Also, I'm not sure I truly like it - it's in the repo but we put it into ignore... maybe I would rather have something like pre-commit or GHA that checks that the Pipfile.lock isn't changed by user?

Reply:
> If there is a more effective way to eliminate these conflicts while maintaining consistency across environments and CI, we are open to hearing it.

maybe just run pip from the makefile, in podman container, in the right python version, and with the various switches for platform, etc.

```
python3.11 -m pipenv lock --extra-pip-args "--only-binary=:all: --platform=manylinux_2_28_x86_64 --python-version=3.11 --implementation=cp"
``` 

stuff like that

Reply:
Yep, basically what the GHA would do, just on your local machine. Still, we wouldn't be sure that user generated these this way maybe... so some GHA gating this on PR would be needed anyway.

Reply:
> so some GHA gating this on PR would be needed anyway

we have already a Jira for this from @paulovmr, needs to be pre-refined, refined, planed, .... let's not get ahead of ourselves https://issues.redhat.com/browse/RHOAIENG-8778

Reply:
Conclusion:
Removed the .gitignore from this implementation as we discussed on the release improvements meeting. 
We will do a better investigation in the future about a better tool to manage the python packages, to avoid several issues that the pipenv cause.